### ISSUE TITLE
Possible reconsideration of some new Unicorn rules

### ISSUE THREAD

**Issue by brettz9** [gajus/eslint-config-canonical#23]
While I see there are good motivations behind the following rules, I think they end up being more of a pain, particularly the first one.

- `unicorn/no-useless-undefined`. One not only has to set `array-callback-return` with `allowImplicit` to `true` as per the docs to make this manageable, arguably further weakening that rule, but one has to disable `consistent-return`, since not allowing `undefined` as a return value leads to inconsistent results if a function can return `undefined` or some other value (e.g., like a polyfill for `Array.prototype.find` to give an example of even a native function that behaves this way).
- `unicorn/prefer-optional-catch-binding` - Sometimes it is nice for the bound variable, even if unused, to indicate the purpose of the error. One can add a comment though admittedly, and this rule probably concerns me less than the first.

I'm also not personally a fan of `unicorn/no-reduce` though I can live with that one too.

**Comment by brettz9** [comment]
Btw, I meant to thank you for the quick merge and release. And I appreciate the idea behind the rules--just feeling esp. the first is a little cumbersome.

**Comment by gajus** [comment]
Thanks. Reasonable.

I am also unsure about `unicorn/prefer-set-has`. Curious to know your thoughts.

**Comment by brettz9** [comment]
Thanks!

Re: `unicorn/prefer-set-has`, on the one hand, I'm not too fond in my own projects of complicating APIs by returning Sets as it may encourage (nor fond of Maps replacing all keyed objects). 

Consumers have to use `[...set.values()]` to be able to get at all the array benefits like array extras or getting serialized results of the contents with `JSON.stringify()`, but I do see the reason for it being good for performance and maybe similar to how `const` is clarifying, it is good to clarify that the contents are unique.

In short, I'm ok with the rule myself, esp. as our use in `eslint-plugin-jsdoc` doesn't call for exposing public APIs.

**Comment by gajus** [comment]
I tried it on a few different codebases, and as you say, I do get the performance benefits, but I dislike complicating the API. If there was an option to give warnings without auto-fixing it, I would be cool with it. But with autofixing it get annoying.

**Comment by brettz9** [comment]
Before I saw your commit, I started applying in `eslint-plugin-jsdoc`. I can revert if you want, but since the current arrays are not inline, I don't think it hurts to define them as sets--even if you no longer want to enforce the rule. But whatever you like.

**Comment by gajus** [comment]
> Before I saw your commit, I started applying in eslint-plugin-jsdoc. I can avoid committing if you want, but since the current arrays are not inline, I don't think it hurts to define them as sets--even if you no longer want to enforce the rule. But whatever you like.

Go for it. I think there is benefit to using sets, it is just that I don't want to enforce auto-fix for it.
